common/concurrent_queue.cpp

```diff
@@ -0,0 +1,107 @@
+#include "concurrent_queue.h"
+
+#include <cassert>
+#include <cstddef>
+#include <mutex>
+#include <utility>
+
+using namespace YACReader;
+
+ConcurrentQueue::ConcurrentQueue(std::size_t threadCount)
+{
+    threads.reserve(threadCount);
+    for (; threadCount != 0; --threadCount)
+        threads.emplace_back(&ConcurrentQueue::nextJob, this);
+}
+
+ConcurrentQueue::~ConcurrentQueue()
+{
+    {
+        std::lock_guard<std::mutex> lock(queueMutex);
+        assert(!bailout);
+        bailout = true;
+    }
+    jobAvailableVar.notify_all();
+
+    for (auto &x : threads)
+        x.join();
+    assert(jobsLeft == _queue.size() && "Only not yet started jobs are left.");
+}
+
+void ConcurrentQueue::enqueue(Job job)
+{
+    {
+        std::lock_guard<std::mutex> lock(jobsLeftMutex);
+        ++jobsLeft;
+    }
+
+    {
+        std::lock_guard<std::mutex> lock(queueMutex);
+        _queue.emplace(std::move(job));
```

Comment:
According to Sean Parent, `std::list` should be used instead of `std::queue` to minimize the locking time with  [`std::list::splice`](https://en.cppreference.com/w/cpp/container/list/splice). See https://www.youtube.com/watch?v=W2tWOdzgXHA&t=45m56s. What do you think?